### PR TITLE
refactor(firebase_database): firebase_database_platform_interface

### PR DESCRIPTION
**Pull request by deepak786**
## Description
1. create separate firebase_database_platform_interface package
2. federate firebase_database plugin

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by deepak786** [comment]
@googlebot I signed it!

**Review by deepak786 — COMMENTED** [review]
if we make this enum (EventType) public (which is not required for now) then need to update the event string in native sides also. Should I make this EvetType type public?

**Comment by deepak786** [comment]
@Salakar I have submitted this pull request which will create a separate firebase_platform_database_reference and federate the firebase_database.
But as per the comment https://github.com/FirebaseExtended/flutterfire/pull/1835#issuecomment-827475385, you also want to improve the integration tests.
Could you please guide or help to improve the integration tests?

**Comment by deepak786** [comment]
@russellwheatley @Salakar

**Comment by deepak786** [comment]
check **all_plugins / analyze (pull_request)** is failing because I have used the path dependency of `firebase_database_platform_interface` to federate the `firebase_database` plugin.

File: packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database.dart [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+  /// The [MethodChannel] used to communicate with the native plugin
+  static const MethodChannel channel =
+      MethodChannel('plugins.flutter.io/firebase_database');
+
+  /// Gets a DatabaseReference for the root of your Firebase Database.
+  @override
+  DatabaseReferencePlatform reference() {
+    return MethodChannelDatabaseReference(
+        database: this, pathComponents: <String>[]);
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database.dart:67]: trailing comma

File: packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        );
+
+  /// Gets a DatabaseReference for the location at the specified relative
+  /// path. The relative path can either be a simple child key (e.g. ‘fred’) or
+  /// a deeper slash-separated path (e.g. ‘fred/name/first’).
+  @override
+  DatabaseReferencePlatform child(String path) {
+    return MethodChannelDatabaseReference(
+        database: database,
+        pathComponents: List<String>.from(pathComponents)
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart:35]: unnecessary `List<String>.from`

File: packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+  /// refers to the root of your Firebase Database, it has no parent, and
+  /// therefore parent() will return null.
+  @override
+  DatabaseReferencePlatform? parent() {
+    if (pathComponents.isEmpty) {
+      return null;
+    }
+    return MethodChannelDatabaseReference(
+      database: database,
+      pathComponents: (List<String>.from(pathComponents))..removeLast(),
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart:49]: unnecessary `List<String>.from`

File: packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  /// Database location represent a list of items.
+  ///
+  /// The unique key generated by childByAutoId: is prefixed with a
+  /// client-generated timestamp so that the resulting list will be
+  /// chronologically-sorted.
+  @override
+  DatabaseReferencePlatform push() {
+    return MethodChannelDatabaseReference(
+      database: database,
+      pathComponents: List<String>.from(pathComponents)
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart:78]: unnecessary `List<String>.from`

File: packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+  /// Performs an optimistic-concurrency transactional update to the data at
+  /// this Firebase Database location.
+  @override
+  Future<TransactionResultPlatform> runTransaction(
+      TransactionHandler transactionHandler,
+      {Duration timeout = const Duration(seconds: 5)}) async {
+    assert(timeout.inMilliseconds > 0,
+        'Transaction timeout must be more than 0 milliseconds.');
+
+    final Completer<TransactionResultPlatform> completer =
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart:180]: drop value type

File: packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    await MethodChannelDatabase.channel.invokeMethod<void>(
+        'DatabaseReference#runTransaction', <String, dynamic>{
+      'app': database.app?.name,
+      'databaseURL': database.databaseURL,
+      'path': path,
+      'transactionKey': transactionKey,
+      'transactionTimeout': timeout.inMilliseconds
+    }).then((dynamic response) {
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_database_reference.dart:209]: trailing comma

File: packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_query.dart [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+          pathComponents: pathComponents,
+        );
+
+  @override
+  Map<String, dynamic> buildArguments() {
+    return Map<String, dynamic>.from(parameters)
+      ..addAll(<String, dynamic>{
+        'path': path,
+      });
+  }
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_query.dart:27]: this override is exactly the same as in the base class

File: packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_query.dart [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+            'app': database.app?.name,
+            'databaseURL': database.databaseURL,
+            'path': path,
+            'parameters': parameters,
+            'eventType': eventType.toString(),
+          },
+        ).then<int>((dynamic result) => result);
+        _handle.then((int handle) {
+          MethodChannelDatabase._observers[handle] = controller;
+        });
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/method_channel/method_channel_query.dart:49]:
```suggestion
        .then((result) {
          MethodChannelDatabase._observers[result!] = controller;
        });
```

File: packages/firebase_database/firebase_database_platform_interface/lib/src/platform_interface/query.dart [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  /// Assigns the proper event type to a stream for [EventPlatform]
+  Stream<EventPlatform> observe(EventType eventType) {
+    throw UnimplementedError('observe() not implemented');
+  }
+
+  Map<String, dynamic> buildArguments() {
+    return Map<String, dynamic>.from(parameters)
+      ..addAll(<String, dynamic>{
+        'path': path,
+      });
```
Comment (lesnitsky) [inline packages/firebase_database/firebase_database_platform_interface/lib/src/platform_interface/query.dart:40]:
```suggestion
    return {
      ...parameters,
      'path': path,
    };
```

**Review by lesnitsky — COMMENTED** [review]
all comments are related to the code which was moved around, not authored by the @deepak786, so not sure if that should be of his concern. @Salakar thoughts?

**Comment by lesnitsky** [comment]
@deepak786 would you be willing to give your consent for co-authoring this? I've fixed my comments myself 😉

**Comment by lesnitsky** [comment]
@googlebot I consent.

**Comment by deepak786** [comment]
@googlebot I consent.

**Comment by deepak786** [comment]
@lesnitsky is that you asked?

**Comment by lesnitsky** [comment]
@deepak786 yes, thanks

**Pull request merged**